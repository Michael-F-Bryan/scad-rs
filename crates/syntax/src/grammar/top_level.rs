use crate::{
    grammar::{expressions, TokenSet},
    parser::{Mark, Parser},
    SyntaxKind::*,
};

const CONTINUE: TokenSet = TokenSet::new([T![;], T!["}"]]);

pub(crate) fn package(p: &mut Parser<'_>) {
    let m = p.start();

    while !p.at(EOF) {
        statement(p);
    }

    p.complete(m, PACKAGE);
}

pub(crate) fn statement(p: &mut Parser<'_>) {
    let m = p.start();

    match p.current() {
        T![include] => include(p, m),
        T![use] => use_(p, m),
        IDENT if p.nth_at(1, T![=]) => {
            assignment(p, m);
            p.expect(T![;]);
        }
        _ => p.error_recover("Expected a statement", m, CONTINUE),
    }
}

pub(crate) fn include(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![include]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, INCLUDE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn use_(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![use]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, USE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn assignment(p: &mut Parser<'_>, m: Mark) {
    p.bump(IDENT);

    p.eat(T![=]);

    expressions::expr(p);
    p.complete(m, ASSIGNMENT);
}

#[cfg(test)]
mod tests {
    use super::*;

    parse_tests! {
        empty_package: package(""),
        include_statement: statement(r#"include "./foo/bar""#),
        use_statement: statement(r#"use "./foo/bar""#),
        assignment_statement: statement("x = 42;"),
        kitchen_sink: package(r#"
            include "foo/bar.scad"
            use "baz.scad"
            a = 42;
            b = (42);
            c = 1 + 1;
        "#),
    }
}
