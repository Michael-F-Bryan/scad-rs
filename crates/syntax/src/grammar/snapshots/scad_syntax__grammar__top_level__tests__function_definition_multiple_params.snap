---
source: crates/syntax/src/grammar/top_level.rs
description: "statement(function x(a, b=2, c) = a+b+c;)"
---
Node(
    NAMED_FUNCTION_DEFINITION@0..30
      FUNCTION_KW@0..8 "function"
      WHITESPACE@8..9 " "
      IDENT@9..10 "x"
      L_PAREN@10..11 "("
      PARAMETERS@11..20
        IDENT@11..12
          IDENT@11..12 "a"
        COMMA@12..13 ","
        ASSIGNMENT@13..17
          WHITESPACE@13..14 " "
          IDENT@14..15 "b"
          EQUALS@15..16 "="
          INTEGER@16..17
            INTEGER@16..17 "2"
        COMMA@17..18 ","
        IDENT@18..20
          WHITESPACE@18..19 " "
          IDENT@19..20 "c"
      R_PAREN@20..21 ")"
      WHITESPACE@21..22 " "
      EQUALS@22..23 "="
      BIN_EXPR@23..29
        LOOKUP_EXPR@23..25
          WHITESPACE@23..24 " "
          IDENT@24..25 "a"
        PLUS@25..26
          PLUS@25..26 "+"
        LOOKUP_EXPR@26..27
          IDENT@26..27 "b"
        PLUS@27..28
          PLUS@27..28 "+"
        LOOKUP_EXPR@28..29
          IDENT@28..29 "c"
      SEMICOLON@29..30 ";"
    ,
)
